lib/elixir/lib/task.ex

```diff
@@ -229,30 +250,35 @@ defmodule Task do
   @doc """
   Starts a process linked to the current process.
 
+  `fun` must be a zero-arity anonymous function.
+
   This is often used to start the process as part of a supervision tree.
   """
   @spec start_link((() -> any)) :: {:ok, pid}
-  def start_link(fun) do
+  def start_link(fun) when is_function(fun, 0) do
     start_link(:erlang, :apply, [fun, []])
   end
 
   @doc """
   Starts a task as part of a supervision tree.
   """
   @spec start_link(module, atom, [term]) :: {:ok, pid}
-  def start_link(mod, fun, args) do
-    Task.Supervised.start_link(get_info(self()), {mod, fun, args})
+  def start_link(module, function_name, args)
+      when is_atom(module) and is_atom(function_name) and is_list(args) do
+    Task.Supervised.start_link(get_info(self()), {module, function_name, args})
   end
 
   @doc """
   Starts a task.
 
+  `fun` must be a zero-arity anonymous function.
+
   This is only used when the task is used for side-effects
   (i.e. no interest in the returned result) and it should not
   be linked to the current process.
   """
   @spec start((() -> any)) :: {:ok, pid}
-  def start(fun) do
+  def start(fun) when is_function(fun, 0) do
```

Comment:
@josevalim, does [this comment](https://github.com/elixir-lang/elixir/pull/8087#discussion_r210400252) also apply in this case?